src/pass/zero_elimination.cc

```diff
@@ -0,0 +1,1719 @@
+      }
+    } else {
+      IRVisitor::Visit(node);
+    }
+  }
+
+  void Visit_(const Variable* op) {
+    CHECK(false) << "This case shouldn't happen";
+  }
+
+  void Visit_(const LetStmt* op) {
+    bound.insert(op->var.get());
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const For* op) {
+    bound.insert(op->loop_var.get());
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Let* op) {
+    bound.insert(op->var.get());
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Reduce* op) {
+    for (const auto& iv : op->axis) {
+      bound.insert(iv->var.get());
+    }
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Store* op) {
+    Visit(op->buffer_var);
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Allocate* op) {
+    Visit(op->buffer_var);
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Free* op) {
+    Visit(op->buffer_var);
+    IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Load* op) {
+    Visit(op->buffer_var);
+    IRVisitor::Visit_(op);
+  }
+};
+
+// Get free variables of an expression
+Array<Var> ExprFreeVars(const Expr& expr) {
+  ExprFreeVarsVisitor visitor;
+  visitor.Visit(expr);
+  return visitor.free_array;
+}
+
+// Clone iter vars and return both the new vars and the substitution from old to new.
+std::pair<Array<IterVar>, std::unordered_map<const Variable*, Expr>> CloneIterVars(
+    const Array<IterVar>& vars) {
+  Array<IterVar> new_vars;
+  std::unordered_map<const Variable*, Expr> vmap;
+  for (const IterVar& iv : vars) {
+    IterVar new_v =
+      IterVarNode::make(iv->dom, iv->var.copy_with_suffix(""),
+          iv->iter_type, iv->thread_tag);
+    new_vars.push_back(new_v);
+    vmap[iv->var.get()] = new_v;
+  }
+  return std::make_pair(std::move(new_vars), std::move(vmap));
+}
+
+// Clone reduction by cloning the axis variables.
+Expr CloneReduction(const Expr& expr) {
+  if (const Reduce* red = expr.as<Reduce>()) {
+    Array<IterVar> new_axis;
+    std::unordered_map<const Variable*, Expr> vmap;
+    std::tie(new_axis, vmap) = CloneIterVars(red->axis);
+
+    Array<Expr> src_with_newaxis;
+    for (const auto& src : red->source) {
+      src_with_newaxis.push_back(Substitute(src, vmap));
+    }
+
+    return Reduce::make(red->combiner, src_with_newaxis,
+        new_axis, Substitute(red->condition, vmap), red->value_index);
+  } else {
+    return expr;
+  }
+}
+
+// Convert an array of itervars to an array of inequalities
+Array<Expr> IterVarsToInequalities(const Array<IterVar>& itervars) {
+  Array<Expr> res;
+  for (const IterVar& v : itervars) {
+    res.push_back(GE::make(v->var, v->dom->min));
+    res.push_back(LT::make(v->var, v->dom->min + v->dom->extent));
+  }
+  return res;
+}
+
+// Convert an array of itervars to a map from vars to ranges
+Map<Var, Range> IterVarsToMap(const Array<IterVar>& itervars) {
+  Map<Var, Range> res;
+  for (const IterVar& v : itervars) {
+    res.Set(v->var, v->dom);
+  }
+  return res;
```

Comment:
Should we `std::move` here and below?

Reply:
I guess no https://stackoverflow.com/a/17473869